js/ui/appDisplay.js

```diff
@@ -1820,12 +1821,96 @@ const AppFolderPopup = new Lang.Class({
+// - createIcon(size) function -> actor
+// - countVisible getter -> boolean
+// - count getter -> number
+const AppIconSource = new Lang.Class({
+    Name: 'AppIconSource',
+
+    _init: function(appIcon) {
+        this._appIcon = appIcon;
+        this._notificationSourceSignalIds = [];
+        this._notificationSource = null;
+
+        let daemon = Main.notificationDaemon.gtk;
+        daemon.connect('new-gtk-notification-source', Lang.bind(this, this._onNewGtkNotificationSource));
+        let id = this._desktopLessAppId(this._appIcon.app.get_id());
+        this._onNewGtkNotificationSource(daemon, id, daemon.getSourceFor(id));
+    },
+
+    createIcon: function(iconSize) {
+        return this._appIcon.app.create_icon_texture(iconSize);
+    },
+
+    get count() {
+        if (this._notificationSource)
+            return this._notificationSource.count;
+        return 0;
+    },
+
+    get countVisible() {
+        return this.count > 0;
+    },
+
+    _desktopLessAppId: function(appId) {
+        if (appId.endsWith('.desktop'))
+            return appId.substring(0, appId.length - '.desktop'.length);
+        return appId;
+    },
+
+    _onNewGtkNotificationSource: function(daemon, appId, source) {
+        let thisAppId = this._desktopLessAppId(this._appIcon.app.get_id());
+        let sourceAppId = this._desktopLessAppId(appId);
+
+        if (thisAppId != sourceAppId)
+            return;
+
+        if (this._notificationSource) {
+            for (let idx in this._notificationSourceSignalIds) {
+                let signalId = this._notificationSourceSignalIds[idx];
+                this._notificationSource.disconnect(signalId);
+            }
+
+            this._notificationSourceSignalIds = [];
+            this._notificationSource = null;
+        }
+
+        let signals = ['count-updated', 'icon-updated'];
+
+        if (source) {
+            this._notificationSource = source;
+            let destroySignalId = this._notificationSource.connect('destroy', Lang.bind(this, function() {
+                this._notificationSourceSignalIds = [];
+                this._notificationSource = null;
+            }));
+            this._notificationSourceSignalIds.push(destroySignalId);
+
+            for (let idx in signals) {
+                let signalName = signals[idx];
+                let signalId = this._notificationSource.connect(signalName, Lang.bind(this, function() {
+                    this.emit(signalName);
+                }));
+                this._notificationSourceSignalIds.push(signalId);
+            }
+        }
+        for (let idx in signals)
+            this.emit(signals[idx]);
+    },
+
+});
+Signals.addSignalMethods(AppIconSource.prototype);
+
 const AppIcon = new Lang.Class({
     Name: 'AppIcon',
     Extends: ViewIcon,
 
     _init : function(app, iconParams, params) {
         this.app = app;
+        this._iconSource = new AppIconSource(this);
+        this._sourceActor = null;
```

Comment:
Hmm, dunno. We don't return `this._sourceActor` from the `_createIcon` function, but rather only `this._sourceActor.actor`. What would happen to `this._sourceActor` if it is not a private member and then gets out of scope? Would it be GC'd or would it survive because the source actor binds itself to some signals in its `_init` function? If the former, then we need to keep it alive by having it as a private member, I think.